pkg/dns/treecache.go

```diff
@@ -0,0 +1,312 @@
+		Entries:    make(map[string]interface{}),
+		m:          &sync.RWMutex{},
+	}
+}
+
+func Deserialize(dir string) (*TreeCache, error) {
+	b, err := ioutil.ReadFile(path.Join(dir, dataFile))
+	if err != nil {
+		return nil, err
+	}
+	var hash []byte
+	hash, err = ioutil.ReadFile(path.Join(dir, crcFile))
+	if err != nil {
+		return nil, err
+	}
+	if !reflect.DeepEqual(hash, getMD5(b)) {
+		return nil, fmt.Errorf("Checksum failed")
+	}
+
+	var cache TreeCache
+	err = json.Unmarshal(b, &cache)
+	if err != nil {
+		return nil, err
+	}
+	cache.m = &sync.RWMutex{}
+	return &cache, nil
+}
+
+func (cache *TreeCache) Serialize(dir string) (string, error) {
+	cache.m.RLock()
+	defer cache.m.RUnlock()
+	b, err := json.Marshal(cache)
+	if err != nil {
+		return "", err
+	}
+
+	if len(dir) == 0 {
+		var prettyJSON bytes.Buffer
+		err = json.Indent(&prettyJSON, b, "", "\t")
+
+		if err != nil {
+			return "", err
+		}
+		return string(prettyJSON.Bytes()), nil
+	}
+	if err := ensureDir(dir, os.FileMode(0755)); err != nil {
+		return "", err
+	}
+	if err := ioutil.WriteFile(path.Join(dir, dataFile), b, 0644); err != nil {
+		return "", err
+	}
+	if err := ioutil.WriteFile(path.Join(dir, crcFile), getMD5(b), 0644); err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+func (cache *TreeCache) SetEntry(key string, val interface{}, path ...string) {
+	cache.m.Lock()
+	defer cache.m.Unlock()
+	node := cache.ensureChildNode(path...)
+	node.Entries[key] = val
+}
+
+func (cache *TreeCache) ReplaceEntries(entries map[string]interface{}, path ...string) {
+	cache.m.Lock()
+	defer cache.m.Unlock()
+	node := cache.ensureChildNode(path...)
+	node.Entries = make(map[string]interface{})
+	for key, val := range entries {
+		node.Entries[key] = val
+	}
+}
+
+func (cache *TreeCache) GetSubCache(path ...string) *TreeCache {
+	childCache := cache
+	for _, subpath := range path {
+		childCache = childCache.ChildNodes[subpath]
+		if childCache == nil {
+			return childCache
+		}
+	}
+	return childCache
+}
+
+func (cache *TreeCache) SetSubCache(key string, subCache *TreeCache, path ...string) {
+	cache.m.Lock()
+	defer cache.m.Unlock()
+	node := cache.ensureChildNode(path...)
+	node.ChildNodes[key] = subCache
+}
+
+func (cache *TreeCache) GetEntry(key string, path ...string) (interface{}, bool) {
+	cache.m.RLock()
+	defer cache.m.RUnlock()
+	childNode := cache.GetSubCache(path...)
+	val, ok := childNode.Entries[key]
+	return val, ok
+}
+
+func (cache *TreeCache) GetValuesForPathWithRegex(path ...string) []interface{} {
+	cache.m.RLock()
+	defer cache.m.RUnlock()
+	retval := []interface{}{}
+	nodesToExplore := []*TreeCache{cache}
+	for idx, subpath := range path {
+		nextNodesToExplore := []*TreeCache{}
+		if idx == len(path)-1 {
+			// if path ends on an entry, instead of a child node, add the entry
+			for _, node := range nodesToExplore {
+				if subpath == "*" || subpath == "any" {
+					nextNodesToExplore = append(nextNodesToExplore, node)
```

Comment:
Isn't this a loop?  I think you want to add all childNodes of node, not node itself.  Consider when path is just `["*"]` - first time through you find the root of your tree, you match "*" so you add the root of the tree back onto the list.  a unit test would have caught that